### ISSUE TITLE
Data.Nat.Primality: add idea of Compositeness

### ISSUE THREAD

**Issue by Taneb** [agda/agda-stdlib#1659]
Currently `Data.Nat.Primality` defines:

```agda
Prime : ℕ → Set
Prime 0             = ⊥
Prime 1             = ⊥
Prime (suc (suc n)) = (i : Fin n) → 2 + toℕ i ∤ 2 + n
```

This is fine, but in many situations a negative proof of `Prime n` is a lot less useful than knowing even one factor that causes it to fail the condition.

I'm proposing to add something like

```agda
Composite : ℕ → Set
Composite 0 = ⊥
Composite 1 = ⊥
Composite (suc (suc n)) = Σ[ i ∈ Fin n ] 2 + toℕ i ∣ 2 + n
```

with a corresponding

```agda
composite? : Decidable Composite
```
implemented similarly to `prime?`, along with proofs relating `Composite` and `Prime` together.

**Comment by JacquesCarette** [comment]
Yes please!

**Comment by MatthewDaggitt** [comment]
Sounds good to me as well!

Proposed minor cosmetic tweak to the definition which should also be propagated through to `Prime` as well:
```
Composite : ℕ → Set
Composite 0 = ⊥
Composite 1 = ⊥
Composite n@(suc (suc _)) = Σ[ i ∈ Fin n ] (toℕ i ∣ n)
```

**Comment by gallais** [comment]
My instinct would be to also implement

```
decide :  Π[ Prime ∪ Composite ]
```

just like we have `decide :  Π[ P ∪ Q ] → Π[ All P ∪ Any Q ]` for lists.

**Comment by Taneb** [comment]
@MatthewDaggitt 
> Sounds good to me as well!
> 
> Proposed minor cosmetic tweak to the definition which should also be propagated through to `Prime` as well:
> 
> ```
> Composite : ℕ → Set
> Composite 0 = ⊥
> Composite 1 = ⊥
> Composite n = Σ[ i ∈ Fin n ] (toℕ i ∣ n)
> ```

Sadly in this case the matching on `suc (suc n)` is necessary. The definition uses the `2 +` to exclude checking whether the prime candidate is a multiple of 1.

**Comment by MatthewDaggitt** [comment]
Ah in which case let's go for:
```
Composite : ℕ → Set
Composite 0 = ⊥
Composite 1 = ⊥
Composite n@(suc (suc _)) = Σ[ i ∈ Fin n ] (toℕ i ∣ n)

```

**Comment by MatthewDaggitt** [comment]
Or am I still getting it wrong? The bracketing is kind of confusing...

**Comment by Taneb** [comment]
> Ah in which case let's go for:
> 
> ```
> Composite : ℕ → Set
> Composite 0               = ⊥
> Composite 1               = ⊥
> Composite n@(suc (suc _)) = Σ[ i ∈ Fin n ] (toℕ i ∣ n)
> ```

This still allows me to use `1 ∣ n` as a proof that n is composite. We have to be careful that we allow neither this nor `n ∣ n` otherwise everything's composite

**Comment by MatthewDaggitt** [comment]
Oh doh, of course. Please ignore me :smile:

**Comment by jamesmckinna** [comment]
Hmmm... (Harrumph? Sigh?)
... consider (nothing to do with `Composite` per se, all for that!): 
* first, deprecating `Prime` in favour of `Irreducible`, adding the correct definition of `Prime`;  
* then, moving both to ... `Algebra...IntegralDomain` under `Ring` ... developing `UniqueFactorisationDomain`, `EuclideanDomain` (where the notions coincide) and show `ℤ` is an ED? 
This second one's a bigger, separate, issue, perhaps http://www.cs.ru.nl/~freek/100/#80 ;-)

**Comment by Taneb** [comment]
I think a definition of prime numbers for doing number theory and stuff with and a definition of irreducible and prime elements of a ring under the Algebra hierarchy can coexist perfectly happily :)

**Comment by JacquesCarette** [comment]
That's actually a really non-trivial 'philosophical' question about library building: should it be bottom-up, where concrete objects are programmed ab initio? Or top-down, where abstraction reigns supreme, and concreteness is obtained by specialization? Or maybe a mixture of both?

The problem one encounters with a lot of top-down is that sets up a huge barrier to entry: if you don't know all that abstract stuff, you can't contribute. That's a bit appealing to part of me that likes preventing ad hoc hacks, but frightening to the part of me who understands that library scale is directly related to community size.

Another huge problem is one of cyclic dependencies and bootstrapping. Math itself is done by a strange mixture of top-down and bottom-up, which doesn't bother mathematicians at all. But makes a real mess of a library.  

It fun to look at the very large Coq and Lean developments to see these kinds of meta-choices.  [Mizar and Isabelle contain way too much duplication and not enough reuse, so they've sidestepped a lot of that pain the cheap way.]  Lean has often chosen the 'high' road, which I find quite interesting.  I have no idea if that's related to being classical or not!

**Comment by jamesmckinna** [comment]
Absolutely, re: @Taneb 's comment.
But even for `ℤ`, the definitions of, and equivalence between, both `Prime` and `Irreducible` would be worth having as a first step. 
Cf. McBride's proof of `Prime 2` in https://link.springer.com/book/10.1007/11542384
Re @JacquesCarette 's point, well I've just been round that loop for PR #1660 with `_^_` in `Data.Integer.Base` and `_ ^ᵗ_` in `Algebra.Definitions.RawSemiring` (where, not entirely (co-)incidentally, `Prime` and `Irreducible` are given the correct definitions!). The hard part is making sure that 'high-style' (top-down Bourbachisme) and 'low-style' (bottom-up) peacefully coexist, but at least correct+consistent naming have their part to play in that...

**Comment by jamesmckinna** [comment]
Ob: Lean. The very definition of 'perfectoid space' https://leanprover-community.github.io/lean-perfectoid-spaces/ is the epitome of 'high-style', so it's not surprising that to get there starting from nothing, the mathematicians using Lean needed to ... make the choices they made. Not sure that classical logic had much to do with it, other than to avoid mathematicians not taking Lean, or the activity of formalisation, seriously...

**Comment by JacquesCarette** [comment]
Peaceful co-existence is indeed a rather delicate business. Seriously under-estimated by those who haven't tried it.

Yes, perfectoid spaces are the epitome of that. But other papers, in the last 2 year's of CPP and the upcoming one, have quite a number of buried design gems. They're consistently along the lines of "We've found that if we design X using this more recent (i.e. from the last 50 years, rather than what's taught in undergrad!) view of math gizmo Y, quite a few things go much more smoothly". And the more recent view is oftentimes more general. Liberating many definitions from the clutches of explicit Nat, Int or Real is often a big win. [My struggles with Multicategory evaporated when I abandoned List and Vec for arguments and went with a general index type instead.]
